app/src/main/java/com/duckduckgo/app/browser/favicon/FaviconManager.kt

```diff
@@ -21,163 +21,212 @@ import android.net.Uri
+    private suspend fun downloadFaviconFor(domain: String): Bitmap? {
+        val faviconUrl = getFaviconUrl(domain) ?: return null
+        val touchFaviconUrl = getTouchFaviconUrl(domain) ?: return null
+        faviconDownloader.getFaviconFromUrl(touchFaviconUrl)?.let {
+            return it
+        } ?: faviconDownloader.getFaviconFromUrl(faviconUrl).let {
+            return it
         }
     }
 
-    private suspend fun loadFaviconToView(file: File, view: ImageView) {
-        faviconDownloader.loadFaviconToView(file, view)
+    private fun getFaviconUrl(domain: String): Uri? {
+        return "https://$domain".toUri().faviconLocation()
     }
 
-    private suspend fun replacePersistedFavicons(icon: Bitmap, domain: String) {
-        if (persistedFaviconsForDomain(domain) > 0) {
-            saveToFile(FAVICON_PERSISTED_DIR, NO_SUBFOLDER, icon, domain)
-        }
+    private fun getTouchFaviconUrl(domain: String): Uri? {
+        return "https://$domain".toUri().touchFaviconLocation()
     }
 
-    private suspend fun getFaviconForDomainOrFallback(directory: String, subFolder: String, domain: String): File? {
-        val cachedFavicon: File? = faviconPersister.faviconFile(directory, subFolder, domain)
+    private suspend fun replacePersistedFavicons(icon: Bitmap, domain: String): File? {
+        return if (persistedFaviconsForDomain(domain) > 0) {
+            faviconPersister.store(FAVICON_PERSISTED_DIR, NO_SUBFOLDER, icon, domain)
+        } else null
+    }
 
-        return if (cachedFavicon == null) {
-            val favicon = downloadFromUrl(domain)
-            if (favicon != null) {
-                saveToFile(directory, subFolder, favicon, domain)
-            } else {
-                null
-            }
+    private suspend fun tryRemoteFallbackFavicon(subFolder: String?, domain: String): File? {
+        val favicon = downloadFaviconFor(domain)
+        return if (favicon != null) {
+            saveFavicon(subFolder, favicon, domain)
         } else {
-            cachedFavicon
+            null
         }
     }
 
-    private suspend fun saveToFile(directory: String, subFolder: String, icon: Bitmap, domain: String): File? {
-        return faviconPersister.store(directory, subFolder, icon, domain)
-    }
-
     private suspend fun persistedFaviconsForDomain(domain: String): Int {
         val query = "%$domain%"
         return withContext(dispatcherProvider.io()) {
             bookmarksDao.bookmarksCountByUrl(query) +
                 locationPermissionsRepository.permissionEntitiesCountByDomain(query) +
-                fireproofWebsiteRepository.fireproofWebsitesCountByDomain(domain)
+                fireproofWebsiteRepository.fireproofWebsitesCountByDomain(domain) +
+                favoritesRepository.favoritesCountByDomain(query)
         }
     }
 
-    private fun extractDomain(url: String): String {
-        return url.toUri().domain() ?: url
+    private fun String.extractDomain(): String? {
```

Comment:
is this the best place to put this Extension function?

Reply:
I think it's ok to keep it here. Since it's private, the logic inside the method already existed on that class but outside the extension function, and we have the context of what that String means.